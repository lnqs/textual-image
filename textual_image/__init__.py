"""Package to render images in Rich and Textual.

Provides Rich Renderables and a Textual Widget to render images in the terminal.
"""
