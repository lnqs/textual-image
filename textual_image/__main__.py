#!/usr/bin/env python

"""Run the textual_image demo."""

import sys
from argparse import ArgumentParser
from importlib.util import find_spec

from textual_image.demo.renderable import RENDERING_METHODS

textual_available = bool(find_spec("textual"))
default_mode = "rich" if find_spec("textual") is None else "textual"

parser = ArgumentParser(description="Demo the capabilities of textual-image")
parser.add_argument("mode", choices=["rich", "textual"], nargs="?", default=default_mode)
parser.add_argument("-m", "--method", choices=RENDERING_METHODS.keys(), default="auto")
arguments = parser.parse_args()

if arguments.mode == "rich":
    from textual_image.demo.renderable import run as run_rich_demo

    run_rich_demo(arguments.method)
elif not textual_available:
    sys.stderr.write(
        "Optional Textual dependency not available. Install this package as `textual-image[textual]` for Textual support."
    )
else:
    from textual_image.demo.widget import run as run_textual_demo

    run_textual_demo(arguments.method)
