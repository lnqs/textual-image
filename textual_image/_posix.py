# pragma: exclude file -- this is platform specific code, either this or _win32.py won't get tested,
# depending on the platform the tests are running

import os
import sys
import termios
import tty
from array import array
from contextlib import contextmanager
from fcntl import ioctl  # type: ignore
from select import select
from typing import Iterator, Tuple


class PosixTerminalError(Exception):
    """Error thrown on failing terminal operations."""

    pass


def get_tiocgwinsz() -> Tuple[int, int, int, int]:
    if not sys.__stdout__:
        raise PosixTerminalError("stdout is closed")

    buf = array("H", [0, 0, 0, 0])
    ioctl(sys.__stdout__, termios.TIOCGWINSZ, buf)  # type: ignore
    rows, columns, screen_width, screen_height = buf
    return rows, columns, screen_width, screen_height


def read(fd: int, length: int, timeout: float | None = None) -> str:
    readable, _, _ = select([fd], [], [], timeout)
    if not readable:
        raise TimeoutError("Timeout waiting for data")
    return os.read(fd, length).decode()


@contextmanager
def capture_mode() -> Iterator[None]:
    if not sys.__stdin__:
        raise PosixTerminalError("stdout is closed")

    stdin = sys.__stdin__.buffer.fileno()
    original_mode = termios.tcgetattr(stdin)  # type: ignore
    tty.setcbreak(stdin, termios.TCSANOW)  # type: ignore

    try:
        yield
    finally:
        termios.tcsetattr(stdin, termios.TCSANOW, original_mode)  # type: ignore
