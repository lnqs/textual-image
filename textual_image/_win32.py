# pragma: exclude file -- this is platform specific code, either this or _posix.py won't get tested,
# depending on the platform the tests are running

import ctypes
import msvcrt
import os
import sys
from contextlib import contextmanager
from ctypes.wintypes import DWORD
from typing import Iterator, Tuple

kernel32 = ctypes.WinDLL("kernel32")  # type: ignore


class Win32TerminalError(Exception):
    """Error thrown on failing terminal operations."""

    pass


def get_tiocgwinsz() -> Tuple[int, int, int, int]:
    raise OSError("TIOCGWINSZ not supported on Windows")


def read(fd: int, length: int, timeout: float | None = None) -> str:
    if timeout is not None:
        wait_for_object(fd, timeout)
    return os.read(fd, length).decode()


@contextmanager
def capture_mode() -> Iterator[None]:
    if not sys.__stdin__:
        raise Win32TerminalError("stdout is closed")

    stdin = sys.stdin.buffer.fileno()

    original_mode = get_console_mode(stdin)
    set_console_mode(stdin, 0x0200)  # ENABLE_VIRTUAL_TERMINAL_INPUT

    try:
        flush(stdin)
        yield
    finally:
        set_console_mode(stdin, original_mode)


def wait_for_object(fd: int, timeout: float) -> None:
    handle = msvcrt.get_osfhandle(fd)  # type: ignore
    ret: int = kernel32.WaitForSingleObject(handle, DWORD(int(timeout * 1000)))
    if ret == 0x00000102:  # WAIT_TIMEOUT
        raise TimeoutError("Timeout waiting for data")


def get_console_mode(fd: int) -> int:
    handle = msvcrt.get_osfhandle(fd)  # type: ignore
    mode = DWORD()
    kernel32.GetConsoleMode(handle, ctypes.byref(mode))
    return mode.value


def set_console_mode(fd: int, mode: int) -> None:
    handle = msvcrt.get_osfhandle(fd)  # type: ignore
    kernel32.SetConsoleMode(handle, mode)


def flush(fd: int) -> None:
    handle = msvcrt.get_osfhandle(fd)  # type: ignore
    kernel32.FlushConsoleInputBuffer(handle)
