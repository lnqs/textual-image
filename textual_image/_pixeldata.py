"""Provides access to pixel data."""

import io
from base64 import b64encode
from contextlib import nullcontext
from pathlib import Path
from typing import ContextManager, Iterable, Iterator, Literal, Tuple

from PIL import Image as PILImage

from textual_image._utils import grouped


def ensure_image(image: str | Path | PILImage.Image) -> ContextManager[PILImage.Image]:
    """Ensures value to be an `PIL.Image.Image`.

    This function accepts either a str or `pathlib.Path` of a path to an image file or a `PIL.Image.Image` instance.
    It returns a context manager that either just provides the passed `PIL.Image.Image` instance or opens the path
    with `PIL` and provides the result.

    Args:
        image: Path to an image file or `PIL.Image.Image` instance.

    Returns:
        A context manager providing a `PIL.Image.Image` instance.
    """
    if isinstance(image, PILImage.Image):
        return nullcontext(image)
    return PILImage.open(image)


class PixelMeta:
    """Provides access to meta information of an image from a path or `PIL.Image.Image` instance."""

    def __init__(self, image: str | Path | PILImage.Image) -> None:
        """Initializes a PixelMeta.

        Args:
            image: Path to an image file or `PIL.Image.Image` instance with the image data to render.
        """
        with ensure_image(image) as opened_image:
            self.width = opened_image.width
            self.height = opened_image.height


class PixelData:
    """Provides access to pixel data from a path or `PIL.Image.Image` instance."""

    def __init__(self, image: str | Path | PILImage.Image, mode: Literal["grayscale", "rgb"] | None = None) -> None:
        """Initializes a PixelData.

        Args:
            image: Path to an image file or `PIL.Image.Image` instance with the image data to render.
            mode: Mode to convert the image into or `None` to keep the original mode.
        """
        with ensure_image(image) as opened_image:
            self._image = opened_image.copy()

        if mode == "grayscale":
            self._image = self._image.convert("L")
        elif mode == "rgb":
            self._image = self._image.convert("RGB")

    @property
    def pil_image(self) -> PILImage.Image:
        """The image as `PILImage.Image`."""
        return self._image

    @property
    def width(self) -> int:
        """The image's width."""
        return self._image.width

    @property
    def height(self) -> int:
        """The image's height."""
        return self._image.height

    def scaled(self, width: int, height: int) -> "PixelData":
        """Returns a scaled copy of this PixelData.

        Args:
            width: Width to scale to.
            height: Height to scale to.

        Returns:
            A new `PixelData` instance of the same image, scaled to (width, height).
        """
        scaled_image = self._image.resize((width, height))
        return PixelData(scaled_image)

    def cropped(self, left: int, top: int, right: int, bottom: int) -> "PixelData":
        """Returns a cropped copy of this PixelData.

        Args:
            left: Left position of the crop rectangle.
            top: Top position of the crop rectangle.
            right: Right position of the crop rectangle.
            bottom: Bottom position of the crop rectangle.

        Returns:
            A new `PixelData` instance of the same image, cropped (left, top, right, bottom).
        """
        cropped_image = self._image.crop((left, top, right, bottom))
        return PixelData(cropped_image)

    def to_base64(self) -> str:
        """Return the pixel data as base64 encoded PNG.

        Returns:
        The image data, as base64 encoded PNG.
        """
        image_buffer = io.BytesIO()
        self._image.save(image_buffer, format="png", compress_level=2)
        return b64encode(image_buffer.getvalue()).decode("ascii")

    def __iter__(self) -> Iterator[Iterable[int | Tuple[int, ...]]]:
        """Iterate the pixel data.

        Returns an `Iterator` over rows of the image.
        The rows are `Iterable`s of pixel values.

        Example:
            Iterate the image data::

            for row in PixelData("image.png"):
                for pixel in row:
                    ...  # Do something with the pixel here

        Returns:
            An `Iterator` over `Iterable`s of pixel values.
        """
        data = self._image.getdata()
        yield from grouped(data, self._image.width)
