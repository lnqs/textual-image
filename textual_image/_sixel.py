from itertools import groupby

from PIL import Image as PILImage

from textual_image._utils import grouped

DCS = "\x1bP"
ST = "\x1b\\"
COLORS = 256


def image_to_sixels(image: PILImage.Image) -> str:
    """Convert an image to Sixels.

    This function convert a `PIL.Image.Image` into the Sixel format.

    Args:
        image: Image to convert to Sixels

    Returns:
        The `image` as Sixel data.
    """
    image = image.convert("RGB").convert("P", palette=PILImage.Palette.ADAPTIVE, colors=COLORS)

    return "".join(
        [
            _get_header(image),
            _get_body(image),
            _get_terminator(),
        ]
    )


def _get_header(image: PILImage.Image) -> str:
    # See <https://vt100.net/docs/vt3xx-gp/chapter14.html#S14.2.1> for details about the parameters.
    sixel_mode = f"{DCS}0;0;0"
    raster_attributes = f'q"1;1;{image.width};{image.height}'

    color_registers = []
    for i, color in enumerate(grouped(image.getpalette() or [], 3)):
        color_str = ";".join(str(int(channel / 256 * 100)) for channel in color)
        color_registers.append(f"#{i};2;{color_str}")

    return f'{sixel_mode}{raster_attributes}{"".join(color_registers)}'


def _get_body(image: PILImage.Image) -> str:
    tokens = []

    for y, row in enumerate(grouped(image.getdata(), image.width)):
        n = 1 << y % 6
        for color, group in groupby(row):
            tokens.append(f"#{color}")

            count = len(list(group))
            if count < 3:
                tokens.append(chr(0x3F + n) * count)
            else:
                tokens.append(f"!{count}{chr(0x3F + n)}")

        tokens.append("-" if n == 32 else "$")

    return "".join(tokens)


def _get_terminator() -> str:
    return ST
