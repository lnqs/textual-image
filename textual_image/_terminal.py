"""Functionality to interact with the terminal."""

import logging
import sys
from contextlib import contextmanager
from types import SimpleNamespace
from typing import Iterator, NamedTuple, cast

# pragma: no cover: start -- platform specific, we always will only execute one branch
if sys.platform == "win32":
    from textual_image._win32 import capture_mode, get_tiocgwinsz, read
else:
    from textual_image._posix import capture_mode, get_tiocgwinsz, read
# pragma: no cover: stop

logger = logging.getLogger(__name__)


class TerminalError(Exception):
    """Error thrown on failing terminal operations."""

    pass


class CellSize(NamedTuple):
    """Size of terminal cells."""

    width: int
    """Width of a terminal cell in pixels."""
    height: int
    """Height of a terminal cell in pixels."""


def get_cell_size() -> CellSize:
    """Get size information from the terminal.

    This function is querying the terminal only once. For any call after the first, a cached result is returned.

    Returns:
        The size information

    """
    if not sys.__stdout__:
        raise TerminalError("stdout is closed")

    if hasattr(get_cell_size, "_result"):
        return cast(CellSize, get_cell_size._result)

    width = 0
    height = 0

    if sys.__stdout__.isatty():
        # Try to get the cell size via ioctl
        try:
            rows, columns, screen_width, screen_height = get_tiocgwinsz()
            width = int(screen_width / columns)
            height = int(screen_height / rows)
        except OSError:
            logger.debug("Failed to get cell size via ioctl, falling back to escape sequence")

    if sys.__stdout__.isatty() and (height == 0 or width == 0):
        # Didn't work, let's try to do it via escape sequence
        try:
            with capture_terminal_response("\x1b[", "t", 0.1) as response:
                sys.__stdout__.write("\x1b[16t")
                sys.__stdout__.flush()

            sequence = response.sequence[len("\x1b[") : -len("t")]
            _, height, width = [int(v) for v in sequence.split(";")]
        except (TerminalError, TimeoutError) as e:
            raise TerminalError("Failed to get cell size") from e

    if not sys.__stdout__.isatty():
        logger.debug("Not connected to a terminal, assuming VT340 sizes")
        width = 10
        height = 20

    cell_size = CellSize(width, height)
    setattr(get_cell_size, "_result", cell_size)
    return cell_size


@contextmanager
def capture_terminal_response(
    start_marker: str, end_marker: str, timeout: float | None = None
) -> Iterator[SimpleNamespace]:
    """Captures a terminal response.

    Captures the terminal's response to an escape sequence.
    This is a bit flaky -- keystrokes during reading the response can lead to false answers.
    Additionally, when the terminal does *not* doesn't send an answer, the first character
    of stdin may get lost as this function reads it to determine if it is the response.
    Anyway, as this is improbable to happen, it should be fine.

    Please not this function will not work anymore once Textual is started. Textual runs a threads to read stdin
    and will grab the response.

    Args:
        start_marker: The start sequence of the expected response
        end_marker: The end sequence of the expected response
        timeout: The number of seconds to wait for the response. None to disable timeout.

    Returns:
        The terminal's response
    """
    if not sys.__stdin__:
        raise TerminalError("stdin is closed")

    response = SimpleNamespace(sequence="")

    stdin = sys.__stdin__.buffer.fileno()

    with capture_mode():
        yield response

        while not response.sequence.endswith(end_marker):
            response.sequence += read(stdin, 1, timeout)

            if not response.sequence.startswith(start_marker[: len(response.sequence)]):
                raise TerminalError("Unexpected response from terminal")
