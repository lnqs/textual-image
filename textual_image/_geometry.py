"""Geometry types and logic."""

from types import NoneType
from typing import NamedTuple, Tuple, cast

from textual_image._terminal import CellSize


class ImageSize(NamedTuple):
    """Image size class capable of parsing size specifications and calculating pixel and cell sizes.

    This class expects the size of the source image as `source_pixel_width` and `source_pixel_height`
    and a target width and height as `width` and `height`.
    `width` and `height` can have multiple forms:

    - If set to an integer, they describe the number of cells to be used when rendering the image.
    - If set to a string ending with `%` (e.g. `80%`) they describe the percentage of available render
      space to be used when rending the image.
    - If set to `auto` the image will be scaled to use as much as possible of the available render space
      while keeping the image's aspect ratio.
    - If set to `None` the original images dimension will be used.
    """

    source_pixel_width: int
    source_pixel_height: int
    width: int | str | None
    height: int | str | None

    def validate(self) -> None:
        """Validate the instance's size specifications.

        Raises:
            ValueError: If `width` or `height` are not valid.
        """
        self.validate_value(self.width)
        self.validate_value(self.height)

    @classmethod
    def validate_value(cls, value: int | str | None) -> None:
        """Validate a single size specification.

        Args:
            value: The size specification.

        Raises:
            ValueError: If `width` or `height` are not valid.
        """
        if isinstance(value, (int, NoneType)) or value == "auto":
            return
        if isinstance(value, str) and value.endswith("%") and value[:-1].isdecimal() and int(value[:-1]) >= 0:
            return
        raise ValueError(f"Invalid size: {value}")

    def get_cell_size(self, max_width: int, max_height: int, terminal_sizes: CellSize) -> Tuple[int, int]:
        """Calculates the render size based on original image size, size specification and terminal cell sizes.

        Args:
            max_width: The maximum available render width in cells.
            max_height: The maximum available render height in cells.
            terminal_sizes: Size information of the terminal.

        Returns:
            The render size in cells as tuple of (width, height).
        """
        self.validate()

        if self.source_pixel_height == 0 or self.source_pixel_width == 0:
            return 0, 0

        width = None
        height = None

        # Just use the original image size for dimensions that're `None`
        if self.width is None:
            width = round(self.source_pixel_width / terminal_sizes.width)
        if self.height is None:
            height = round(self.source_pixel_height / terminal_sizes.height)

        # Explicitly given cell sizes
        if isinstance(self.width, int):
            width = self.width
        if isinstance(self.height, int):
            height = self.height

        # Percentage of available space
        if isinstance(self.width, str) and self.width.endswith("%"):
            width = round(max_width / 100 * int(self.width[:-1]))
        if isinstance(self.height, str) and self.height.endswith("%"):
            height = round(max_height / 100 * int(self.height[:-1]))

        # If `width` is `auto` we use the `max_width` if we don't have a height specified,
        # but keep the ratio if `height` was set
        if self.width == "auto" and self.height is not None:
            ratio = self.source_pixel_width / self.source_pixel_height
            scaled_pixel_height: float = (height or max_height) * terminal_sizes.height
            scaled_pixel_width = scaled_pixel_height * ratio
            width = min(round(scaled_pixel_width / terminal_sizes.width), max_width)
        elif self.width == "auto":
            width = max_width

        # Same for height
        if self.height == "auto" and self.width is not None:
            ratio = self.source_pixel_width / self.source_pixel_height
            scaled_pixel_width = cast(int, width) * terminal_sizes.width
            scaled_pixel_height = scaled_pixel_width / ratio
            height = round(scaled_pixel_height / terminal_sizes.height)
        elif self.height == "auto":
            height = max_height

        # By deriving `height` from `width` in the `auto` case, `height` may exceed `max_height`,
        # which shouldn't happen when set to `auto`. Fix this my scaling down both `height` and `width`.`
        if self.width == "auto" and self.height == "auto" and cast(int, height) > max_height:
            height = max_height
            ratio = self.source_pixel_width / self.source_pixel_height
            scaled_pixel_height = height * terminal_sizes.height
            scaled_pixel_width = scaled_pixel_height * ratio
            width = round(scaled_pixel_width / terminal_sizes.width)

        return cast(Tuple[int, int], (width, height))

    def get_pixel_size(self, max_width: int, max_height: int, terminal_sizes: CellSize) -> Tuple[int, int]:
        """Calculates the pixel size to use for rendering based on original image size, size specification and terminal cell sizes.

        Args:
            max_width: The maximum available render width in cells.
            max_height: The maximum available render height in cells.
            terminal_sizes: Size information of the terminal.

        Returns:
            The render size in pixels as tuple of (width, height).
        """
        width, height = self.get_cell_size(max_width, max_height, terminal_sizes)
        return width * terminal_sizes.width, height * terminal_sizes.height
