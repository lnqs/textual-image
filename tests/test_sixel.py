from PIL import Image as PILImage
from syrupy.assertion import SnapshotAssertion

from tests.data import TEST_IMAGE
from textual_image._sixel import image_to_sixels


def test_image_to_sixels(snapshot: SnapshotAssertion) -> None:
    with PILImage.open(TEST_IMAGE) as image:
        image = image.resize((16, 16))

    assert image_to_sixels(image) == snapshot
