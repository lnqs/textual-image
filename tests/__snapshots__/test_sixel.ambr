# serializer version: 1
# name: test_image_to_sixels
  '\x1bP0;0;0q"1;1;16;16#0;2;78;73;72#1;2;78;73;67#2;2;83;65;49#3;2;91;62;49#4;2;73;71;73#5;2;58;66;79#6;2;91;62;48#7;2;80;59;49#8;2;64;60;61#9;2;61;56;50#10;2;82;55;44#11;2;82;55;41#12;2;83;53;43#13;2;85;54;37#14;2;82;54;32#15;2;71;55;43#16;2;62;54;58#17;2;55;53;55#18;2;81;51;39#19;2;82;50;35#20;2;82;49;38#21;2;83;48;33#22;2;83;48;38#23;2;78;48;36#24;2;82;47;34#25;2;80;46;35#26;2;75;46;31#27;2;72;49;39#28;2;72;49;38#29;2;64;46;45#30;2;62;46;40#31;2;61;48;41#32;2;53;50;51#33;2;53;48;53#34;2;46;60;83#35;2;45;58;81#36;2;45;58;82#37;2;44;57;80#38;2;48;56;71#39;2;42;56;78#40;2;42;56;78#41;2;41;56;79#42;2;43;55;78#43;2;42;55;79#44;2;40;55;79#45;2;41;55;78#46;2;41;55;78#47;2;41;55;78#48;2;41;55;77#49;2;39;55;78#50;2;41;54;78#51;2;39;54;78#52;2;39;53;78#53;2;39;54;78#54;2;51;48;55#55;2;51;47;46#56;2;35;51;76#57;2;35;51;75#58;2;35;51;75#59;2;35;50;75#60;2;37;48;67#61;2;32;48;75#62;2;32;47;73#63;2;31;46;73#64;2;31;46;73#65;2;73;43;28#66;2;69;45;35#67;2;68;41;32#68;2;70;40;26#69;2;68;39;26#70;2;67;40;30#71;2;62;42;42#72;2;65;42;33#73;2;57;45;39#74;2;65;39;26#75;2;52;39;38#76;2;65;37;26#77;2;64;35;25#78;2;69;36;13#79;2;61;37;22#80;2;58;37;29#81;2;51;31;39#82;2;58;30;18#83;2;46;43;45#84;2;46;42;44#85;2;50;39;33#86;2;39;42;56#87;2;41;39;48#88;2;45;38;40#89;2;44;37;39#90;2;41;37;41#91;2;39;35;41#92;2;50;32;40#93;2;39;33;36#94;2;38;35;43#95;2;38;32;26#96;2;43;29;23#97;2;44;29;22#98;2;41;27;23#99;2;30;44;71#100;2;28;43;71#101;2;27;43;71#102;2;27;43;70#103;2;27;42;71#104;2;27;42;70#105;2;27;41;70#106;2;26;41;69#107;2;25;41;69#108;2;25;41;69#109;2;31;39;57#110;2;26;39;66#111;2;25;40;69#112;2;25;40;68#113;2;25;39;69#114;2;25;39;68#115;2;30;38;58#116;2;23;39;68#117;2;33;33;35#118;2;34;33;41#119;2;26;32;46#120;2;33;32;42#121;2;32;31;42#122;2;22;31;49#123;2;35;30;32#124;2;35;30;26#125;2;33;28;26#126;2;32;30;42#127;2;29;28;31#128;2;32;28;23#129;2;39;25;21#130;2;39;23;19#131;2;38;23;21#132;2;47;20;12#133;2;34;23;19#134;2;29;25;24#135;2;30;17;10#136;2;28;27;28#137;2;26;26;28#138;2;27;25;35#139;2;28;21;23#140;2;26;19;13#141;2;25;23;25#142;2;25;23;24#143;2;25;22;25#144;2;18;18;33#145;2;21;17;21#146;2;23;17;24#147;2;17;17;25#148;2;17;17;34#149;2;61;16;16#150;2;50;10;11#151;2;45;16;19#152;2;34;3;8#153;2;33;4;5#154;2;25;16;14#155;2;25;14;15#156;2;25;5;6#157;2;21;16;23#158;2;22;16;14#159;2;19;16;13#160;2;16;14;17#161;2;20;12;9#162;2;16;12;17#163;2;14;19;31#164;2;14;15;33#165;2;16;14;16#166;2;15;14;28#167;2;13;14;21#168;2;14;13;29#169;2;14;13;27#170;2;13;12;26#171;2;14;12;24#172;2;10;11;31#173;2;12;10;19#174;2;10;11;17#175;2;8;11;17#176;2;14;12;16#177;2;13;10;10#178;2;12;10;12#179;2;14;10;12#180;2;10;9;19#181;2;10;10;23#182;2;10;10;12#183;2;8;9;26#184;2;9;8;28#185;2;8;7;13#186;2;10;7;12#187;2;11;7;12#188;2;11;10;11#189;2;11;10;11#190;2;10;8;8#191;2;11;8;10#192;2;16;7;10#193;2;8;6;6#194;2;8;9;28#195;2;7;8;26#196;2;6;6;25#197;2;6;6;25#198;2;6;6;15#199;2;8;8;11#200;2;7;7;12#201;2;7;7;12#202;2;8;7;10#203;2;7;7;9#204;2;6;7;9#205;2;7;6;9#206;2;7;6;9#207;2;5;5;21#208;2;5;5;20#209;2;6;5;15#210;2;8;5;13#211;2;6;5;10#212;2;5;5;9#213;2;5;6;10#214;2;8;7;8#215;2;6;6;8#216;2;6;6;7#217;2;5;5;8#218;2;5;5;8#219;2;5;5;8#220;2;5;5;8#221;2;5;5;6#222;2;5;5;7#223;2;5;5;6#224;2;5;5;6#225;2;5;5;6#226;2;5;5;5#227;2;4;5;7#228;2;3;5;5#229;2;7;3;14#230;2;5;4;8#231;2;5;4;7#232;2;4;4;6#233;2;4;4;5#234;2;4;3;5#235;2;5;4;5#236;2;5;3;5#237;2;4;4;5#238;2;4;3;5#239;2;3;4;5#240;2;3;3;6#241;2;3;3;6#242;2;3;3;4#243;2;2;3;5#244;2;2;2;5#245;2;2;2;4#246;2;2;2;5#247;2;2;1;3#197@#195@#172@#148@#89@#157@#54@#86@#115@#110@#116@#112@#114@#113@#111@#99@$#121A#144A#164A#183A#146A#93A#55A#9A#1A#32A#5A#103A#107A#108A#106A#105A$#126C#120C#87C#91C#139C#160C#159C#128C#125C#95C#124C#122C#101C#104C#102C#100C$#184G#196G#194G#170G#158G#193G#133G#98G#129G#96G#161G#163G#61G#62G#63G#64G$#33O#94O#168O#181O#154O#97O#66O#10O#3O#67O#80O#118O#56O#58O#57O#59O$#166_#138_#169_#173_#130_#76_#77_#18_#6_#23_#26_#29_#53_#49_#51_#52_-#171@#208@#207@#147@#15@#11@#19@#20@#12@#25@#13@#16@#44@#39@#45@#47@$#123A#180A#229A#71A#2A#79A#65A#24A#22A#21A#28A#38A#41A#40A#46A#50A$#209C#210C#152C#149C#14C#78C#74C#69C#70C#68C#88C#48C#34C#35C#37C#43C$#198G#75G#150G#153G#135G#177G#30G#82G#132G#27G#142G#175G#119G#60G#42G#36G$#145O#134O#214O#239O#243O#244O#8O#7O#72O#0O#182O#232O#211O#185O#167O#109O$#73_#236_#233_#237__#245_#127_#117_#136_#137_#216_#199_#200_#201_#230_#174_-#31@#235@#188@#165@#189@#222@#246@#17@#4@#206@#204@#186@#187@#213@#220@#231@$#85A#247A#141A#83A#84A#178A#240A#90A#143A#212A#155A#81A#92A#151A#191A#218A$#131C#242C#238C#234C#221C#224C#225C#241C#227C#203C#179C#162C#176C#192C#215C#217C$#156G#228G#226G#223!3G#233G#140G#190G#204G#203G#202G#205G#219G#227GG$\x1b\\'
# ---
